lxd/cluster/gateway.go

```diff
@@ -666,7 +666,9 @@ func (g *Gateway) LeaderAddress() (string, error) {
 			return "", err
 		}
 		setDqliteVersionHeader(request)
-		request = request.WithContext(ctx)
+		rqCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
```

Comment:
I wonder also if it wouldn't be better to not use the parent `ctx` here, so that it is more predictable that all of the members are checked (as even with a 1s timeout the total of 5s budget could still be exceeded if there were a lot of members).

Reply:
Also it seems like the line `leader, err := client.Leader(context.Background())` would benefit from being passed the 5s ctx also.

Reply:
So like this:

```go
	// If this is a voter node, return the address of the current leader, or wait a bit until one is elected.
	if g.server != nil && g.info.Role == db.RaftVoter {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()

		for {
			client, err := g.getClient()
			if err != nil {
				return "", fmt.Errorf("Failed to get dqlite client: %w", err)
			}

			leader, err := client.Leader(ctx)
			if err != nil {
				client.Close()
				return "", fmt.Errorf("Failed to get leader address: %w", err)
			}

			if leader != nil && leader.Address != "" {
				client.Close()
				return leader.Address, nil
			}
			client.Close()

			select {
			case <-ctx.Done():
				return "", ctx.Err()
			case <-time.After(time.Second):
				continue
			}
		}
	}
```

Reply:
Also `ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)` should probably be `ctx, cancel := context.WithTimeout(g.ctx, 5*time.Second)` so that if the `Gateway.Kill()` function is called that any ongoing requests be cancelled.